### PR TITLE
Fix: dotnet-new3 doesn't respect DOTNET_CLI_UI_LANGUAGE env variable

### PR DESCRIPTION
**Pull request by bozturkMSFT**
### Problem
Fixes https://github.com/dotnet/templating/issues/2754

### Solution
DOTNET_CLI_UI_LANGUAGE env variable is now used to decide on the UI language.
Removed --locale option.

### Checks:
- [x] Added unit tests
- [x] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

**Review by DavidKarlas — APPROVED** [review]
I'm not thrilled about order of class members changes in New3Command.cs, since it will make discovery merging harder

File: src/dotnet-new3/Program.cs [hunk 1]
```diff
@@ -14,7 +14,7 @@
 using Microsoft.TemplateEngine.Cli.PostActionProcessors;
 using Microsoft.TemplateSearch.Common.TemplateUpdate;
 
-[assembly:InternalsVisibleTo("dotnet-new3.UnitTests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
+[assembly: InternalsVisibleTo("dotnet-new3.UnitTests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
```
Comment (vlada-shubina) [inline src/dotnet-new3/Program.cs:22]: not your change - just wondering whether we need it

File: test/dotnet-new3.UnitTests/Sdk/SdkCommandSpec.cs [hunk 2]
```diff
@@ -1,22 +1,30 @@
 ﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+#nullable enable
```
Comment (vlada-shubina) [inline test/dotnet-new3.UnitTests/Sdk/SdkCommandSpec.cs:4]: These classes were copied from dotnet/sdk - just wondering if we want to keep modifying them - it makes them diverge from original and if we need to update them from dotnet/sdk we probably have spend time to merge.

File: src/dotnet-new3/Program.cs [hunk 3]
```diff
@@ -147,5 +150,44 @@ private static void FirstRun(IEngineEnvironmentSettings environmentSettings, IIn
                 paths.Copy(paths.Global.DefaultInstallTemplateSearchData, paths.User.NuGetScrapedTemplateSearchFile);
             }
         }
+
```
[the remark this answers is not in the document]
Comment (DavidKarlas) [reply src/dotnet-new3/Program.cs:162]: This is actually copied from SDK to get `new3` as close as possible to `new` https://github.com/dotnet/sdk/blob/52f079f4/src/Cli/dotnet/UILanguageOverride.cs#L38

File: src/dotnet-new3/Program.cs [hunk 4]
```diff
@@ -147,5 +150,44 @@ private static void FirstRun(IEngineEnvironmentSettings environmentSettings, IIn
...
+        private static void ConfigureLocale()
+        {
+            CultureInfo selectedCulture = null;
+
+            // VSLANG=<lcid> is set by VS and we respect that as well so that we will respect the VS
+            // language preference if we're invoked by VS.
+            string vsLang = Environment.GetEnvironmentVariable(VSLANG);
```
[the remark this answers is not in the document]
Comment (DavidKarlas) [reply src/dotnet-new3/Program.cs]: Yes... https://github.com/dotnet/sdk/blob/52f079f4/src/Cli/dotnet/UILanguageOverride.cs#L38

File: src/dotnet-new3/Program.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bozturkMSFT) [reply src/dotnet-new3/Program.cs:162]: Like David said, I copied that part from CLI. So the behavior should be identical.

File: test/dotnet-new3.UnitTests/Sdk/SdkCommandSpec.cs (hunk 2 above)
Comment (bozturkMSFT) [reply test/dotnet-new3.UnitTests/Sdk/SdkCommandSpec.cs:4]: Sounds like a valid concern to me. I will revert this.

File: src/dotnet-new3/Program.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/dotnet-new3/Program.cs:162]: I think we need to double check if it works as intended after this commit is inserted to SDK, or have you already checked?

[the remark this answers is not in the document]
Comment (DavidKarlas) [reply src/dotnet-new3/Program.cs:162]: Which commit? This change is in `dotnet-new3`, code I linked is already in their master for long time.

**Comment by bozturkMSFT** [comment]
Reverted styling changes to New3Command as well as changes to files directly copied from CLI

File: src/dotnet-new3/Program.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/dotnet-new3/Program.cs:162]: But is it already working for dotnet new built in to CLI? I've tried to set `$env:DOTNET_CLI_UI_LANGUAGE = 'de-DE'` for 5.0.100, and run some commands - nothing gets translated; however I expect the restore messages gets translated (they have localization).

Probably that's due to https://github.com/dotnet/sdk/blob/a50a304d00533ff118e277112ab8cd31c60d17b6/src/Cli/dotnet/commands/dotnet-new/NewCommandShim.cs#L84 uses `CultureInfo.CurrentCulture.Name,`, dotnet new 3 uses `CultureInfo.CurrentUICulture.Name` cause the the code in https://github.com/dotnet/sdk/blob/52f079f4/src/Cli/dotnet/UILanguageOverride.cs#L38
 modifies UI language. 

I think we would need an additional PR to SDK anyway to get it worked, once this is merged and inserted we can test it to see whether it works from SDK CLI.

Also should we set env variable like SDK does, so the underlying processes are translated? 
https://github.com/dotnet/sdk/blob/52f079f4ea2c1165c12e7cda2aa8cf223c2a1cbc/src/Cli/dotnet/UILanguageOverride.cs#L21

File: src/dotnet-new3/Program.cs [hunk 5]
```diff
@@ -66,13 +73,19 @@ private static DefaultTemplateEngineHost CreateHost(bool emitTimings)
...
+                // for assembly: Microsoft.TemplateEngine.Cli
+                typeof(DotnetRestorePostActionProcessor).GetTypeInfo().Assembly,
+                // for assembly: Microsoft.TemplateSearch.Common
+                typeof(NupkgUpdater).GetTypeInfo().Assembly
             });
 
-            DefaultTemplateEngineHost host = new DefaultTemplateEngineHost(HostIdentifier, HostVersion, CultureInfo.CurrentCulture.Name, preferences, builtIns, new[] { "dotnetcli" });
+            ConfigureLocale();
+
+            DefaultTemplateEngineHost host = new DefaultTemplateEngineHost(HostIdentifier, HostVersion, CultureInfo.CurrentUICulture.Name, preferences, builtIns, new[] { "dotnetcli" });
```
Comment (vlada-shubina) [inline src/dotnet-new3/Program.cs:88]: `CultureInfo.CurrentUICulture.Name` is set to `Locale` property of the cost (read-write property). Should we change this property to be read only and return `CultureInfo.CurrentUICulture.Name`. I'm a bit concerned that we have multiple sources for locale: `CultureInfo.CurrentUICulture.Name`, `CultureInfo.CurrentCulture.Name`  and `ITemplateEngineHost.Locale` and they might get out-of sync.

Comment (vlada-shubina) [reply src/dotnet-new3/Program.cs:88]: Or we can simply remove Locale property, not sure what's the benefit of it.

File: src/dotnet-new3/Program.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bozturkMSFT) [reply src/dotnet-new3/Program.cs:162]: We had already created a PR regarding `CurrentUICulture` here: https://github.com/dotnet/sdk/pull/16017 and it is now merged.

Though I'm not sure, `FlowOverrideToChildProcesses` looks like something we need here.

**Review by vlada-shubina — APPROVED** [review]
Not high priority: think of if we should keep ITemplateEngineHost.Locale / UpdateLocale as it is now as I see it as a possible source of mistakes for those who implements the host (remaining comment).

Example: 
- UI culture is de-DE
- host by mistake is passing en-US to host constructor 
- in this case UI will be de-DE localized, but not the template.json elements.

Also we need to check what VS is passing as locale - might be same issue as with SDK.

File: src/dotnet-new3/Program.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bozturkMSFT) [reply src/dotnet-new3/Program.cs:163]: Added `FlowOverrideToChildProcessees` (renamed to `ConfigureLocaleForChildProcesses`).

File: src/dotnet-new3/Program.cs (hunk 1 above)
Comment (bozturkMSFT) [reply src/dotnet-new3/Program.cs:22]: Code seem to build without this. We will probably keep these and add even more when we are fixing the accessibility of types, but removing for the time being.

File: src/dotnet-new3/Program.cs (hunk 5 above)
Comment (bozturkMSFT) [reply src/dotnet-new3/Program.cs]: Locale was always set to either `CultureInfo.CurrentCulture` or to "en-US" (mostly in tests). I don't see much point either. So I removed the locale from the host in the last commit. Please review with care as this maybe an impactful change.

**Comment by DavidKarlas** [comment]
I wonder if we even need to pass CultureInfo around, or we can just rely on `CultureInfo.CurrentUICulture` of current thread...

**Comment by bozturkMSFT** [comment]
One of the places where the locale is passed around is `TemplateCache.WriteTemplateCaches`. I think this is one of the examples of where passing it makes sense. Because we are not always using the `CultureInfo.CurrentUICulture` here and the culture depends on the languages supported by the localized templates.

https://github.com/dotnet/templating/blob/master/src/Microsoft.TemplateEngine.Edge/Settings/TemplateCache.cs#L181

File: src/Microsoft.TemplateEngine.Abstractions/ITemplateEngineHost.cs [hunk 6]
```diff
@@ -18,8 +18,6 @@ public interface ITemplateEngineHost
 
         void LogTiming(string label, TimeSpan duration, int depth);
 
-        string Locale { get; }
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Abstractions/ITemplateEngineHost.cs:21]: just a thought: should we rather keep the property but rather force it to be CultureInfo.CurrentUICulture.Name and UpdateLocale to update UICulture when changed in our implementations of hosts?

random idea: some author host might want to implement Locale differently from CultureInfo.CurrentUICulture and this way they will still be able to do it. 
Also I'm not fan of Edge calling CultureInfo for file naming determining.

Comment (bozturkMSFT) [reply src/Microsoft.TemplateEngine.Abstractions/ITemplateEngineHost.cs:21]: Changing the `CurrentUICulture` inside `UpdateLocale` might have undesired effects. The caller will think that it is changing the locale of the host, but it is changing it for the whole thread. Someone may create 2 different hosts and give them different locales, which again won't work.

For your random idea: Localized resources in .resx will still use CurrentUICulture by default. If someone overrides the default locale, resource managers will need to be updated as well. For me, it sounds simple to have everything get locale from CurrentUICulture.

What part of the code are you talking about with "Edge calling CultureInfo"? I'm open to suggestions.

**Pull request merged**